File: Python/Leetcode75/Easy/BitManipulation/Q338-CountingBits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        return [bin(i).count('1') for i in range(n + 1)]
    

    def countBitsOptimized(self, n: int) -> List[int]:
        if n == 0:
            return [0]
        elif n == 1:
            return [0, 1]

        ans = [0, 1]
        for i in range(2, n + 1):
            if i % 2 == 0:
                ans.append(ans[i // 2])
            else:
                ans.append(ans[i // 2] + 1)
        return ans

File: cpp/Leetcode/BitOperation/Q338-CountingBits.cpp
#include <vector>

class Solution {
public:
  std::vector<int> countBits(int n) {
    std::vector<int> ans(n + 1);

    for (int i = 0; i <= n; ++i) {
      ans[i] = bitCount(i);
    }

    return ans;
  }

  int bitCount(int n) {
    int count = 0;

    while (n > 0) {
      if (n % 2 == 1) {
        ++count;
      }
      n /= 2;
    }

    return count;
  }
};

class SolutionDP {
public:
  std::vector<int> countBits(int n) {
    std::vector<int> ans(n + 1, 0);

    int sub = 1;
    for (int i = 1; i <= n; ++i) {
      if (sub * 2 == i)
        sub = i;
      ans[i] = ans[i - sub] + 1;
    }

    return ans;
  }
};